src/Microsoft.OData.Client/Serialization/HttpClientRequestMessage.cs

```diff
@@ -0,0 +1,530 @@
+        public override IODataResponseMessage EndGetResponse(IAsyncResult asyncResult)
+        {
+            return UnwrapAggregateException(() =>
+            {
+                HttpResponseMessage result = ((Task<HttpResponseMessage>)asyncResult).Result;
+                return ConvertHttpWebResponse(result);
+            });
+        }
+
+        /// <summary>
+        /// Returns a response from an Internet resource.
+        /// </summary>
+        /// <returns>A System.Net.HttpResponseMessage that contains the response from the Internet resource.</returns>
+        public override IODataResponseMessage GetResponse()
+        {
+            return UnwrapAggregateException(() =>
+            {
+                Task<HttpResponseMessage> send = CreateSendTask();
+                send.Wait();
+                _httpResponseMessage = send.Result;
+
+                return ConvertHttpWebResponse(send.Result);
+            });
+        }
+
+        private Task<HttpResponseMessage> CreateSendTask()
+        {
+            // Only set the message content if the stream has been written to, otherwise
+            // HttpClient will complain if it's a GET request.
+            Byte[] messageContent = _messageStream.ToArray();
+            if (messageContent.Length > 0)
+            {
+                _requestMessage.Content = new ByteArrayContent(messageContent);
+
+                _messageStream.Dispose();
+
+                // Apply cached "Content" header values
+                foreach (KeyValuePair<string, string> contentHeader in _contentHeaderValueCache)
+                {
+                    _requestMessage.Content.Headers.Add(contentHeader.Key, contentHeader.Value);
+                }
+            }
+
+            _requestMessage.Method = new HttpMethod(_effectiveHttpMethod);
+
+            return _client.SendAsync(_requestMessage);
+        }
+
+        private static IDictionary<string, string> HttpHeadersToStringDictionary(HttpHeaders headers)
+        {
+            return headers.ToDictionary((h1) => h1.Key, (h2) => string.Join(",", h2.Value));
+        }
+
+        /// <summary> Returns the underlying HttpRequestMessage </summary>
+        internal HttpRequestMessage HttpRequestMessage
+        {
+            get
+            {
+                return _requestMessage;
+            }
+        }
+
+        private static HttpWebResponseMessage ConvertHttpWebResponse(HttpResponseMessage response)
+        {
+            IEnumerable<KeyValuePair<string, string>> allHeaders = HttpHeadersToStringDictionary(response.Headers).Concat(HttpHeadersToStringDictionary(response.Content.Headers));
+            return new HttpWebResponseMessage(
+                allHeaders.ToDictionary((h1) => h1.Key, (h2) => h2.Value),
+                (int)response.StatusCode,
+                () =>
+                {
+                    Task<Stream> task = response.Content.ReadAsStreamAsync();
+                    task.Wait();

```

Comment:
Maybe lets not have this unless if it is necessary as this may cause some issues with some client due to blocking semantics of wait.

Reply:
This is necessary here. I'll like to know though how to go about it if you have different thoughts.

Reply:
@marabooy  For now we will just have to use `.Wait()` until we have a method that returns a Task so that we can use `await`.